### PR TITLE
[react-dom] Refactor event priority handling to its own module

### PR DESCRIPTION
**Pull request by trueadm**
The team have mentioned a bunch of times that it felt wrong that DOM Event Priority handling was all coupled with the SimpleEventPlugin, especially as might want to extend on the priorities with newer features and refactors in the future.

This PR breaks out the priority logic into a dedicated module and makes uses of Maps for faster lookups (I benchmarked before and after and Maps were consistently faster than object property lookups in Chrome and Safari, neutral in Firefox). We should also get a slight code size saving by some of the code golfing that I did whilst refactoring the code.

### CODE REVIEW

File: packages/react-dom/src/events/DOMEventPriority.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ * @flow
+ */
+
+import type {EventPriority} from 'shared/ReactTypes';
+import type {TopLevelType} from 'legacy-events/TopLevelEventTypes';
+
+import * as DOMTopLevelEventTypes from './DOMTopLevelEventTypes';
+import {DiscreteEvent, UserBlockingEvent, ContinuousEvent} from 'shared/ReactTypes';
+
+// We use a Set here rather than an Array or Object map.
```
Comment (philipp-spiess) [inline packages/react-dom/src/events/DOMEventPriority.js]: Out of curiosity: Have you also tried a switch statement?

Comment (trueadm) [reply packages/react-dom/src/events/DOMEventPriority.js]: I did and it was even faster, but not by much. The issue with switches was that it was also quite a bit more code size comared to before.

**Comment by trueadm** [comment]
I did some benchmarking on FF and Safari and it looks like using multiple Sets is not as optimal as Chrome's V8. I've changed my approach to use Maps and moved things around so we don't do multiple lookups – this is now consistently faster in Chrome and Safari, neutral in Firefox.

File: packages/react-dom/src/events/DOMEventProperties.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+} from 'legacy-events/TopLevelEventTypes';
+import type {DispatchConfig} from 'legacy-events/ReactSyntheticEventType';
+
+import * as DOMTopLevelEventTypes from './DOMTopLevelEventTypes';
+import {
+  DiscreteEvent,
+  UserBlockingEvent,
+  ContinuousEvent,
+} from 'shared/ReactTypes';
+import {simpleEventPluiginEventTypes} from './SimpleEventPlugin';
```
Comment (necolas) [inline packages/react-dom/src/events/DOMEventProperties.js]: Typo: simpleEvent**Pluigin**EventTypes

But this file now imports the `SimpleEventPlugin` and creates a cyclic dependency between that and `DOMEventProperties`

File: packages/react-dom/src/events/DOMEventProperties.js [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+ *   [TOP_ABORT, { sameConfig }],
+ * ]);
+ */
+
+function processTopEventTypesByPriority(
+  eventTypes: Array<DOMTopLevelEventType | string>,
+  priority: EventPriority,
+): void {
+  // As the event types are in pairs of two, we need to iterate
+  // through in twos.
```
Comment (philipp-spiess) [inline packages/react-dom/src/events/DOMEventProperties.js]: 😮 neat!

Comment (necolas) [reply packages/react-dom/src/events/DOMEventProperties.js]: eek though

File: packages/react-dom/src/events/DOMEventProperties.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  ContinuousEvent,
+} from 'shared/ReactTypes';
+
+// Needed for SimpleEventPlugin, rather than
+// do it in two places, which duplicates logic
+// and increases the bundle size, we do it all
+// here once. If we remove or refactor the
+// SimpleEventPlugin, we should also remove or
+// update the below line.
+export const simpleEventPluiginEventTypes = {};
```
Comment (necolas) [inline packages/react-dom/src/events/DOMEventProperties.js]: Typo here, should be "Plugin"

**Pull request merged**